src/continuous_agg.c

```diff
@@ -259,59 +358,165 @@ ts_continuous_agg_get_all_caggs_info(int32 raw_hypertable_id)
 	List *caggs = ts_continuous_aggs_find_by_raw_table_id(raw_hypertable_id);
 	ListCell *lc;
 	Datum bucket_width;
+	ContinuousAggsBucketFunction **bucket_function;
 
 	all_caggs_info.bucket_widths = NIL;
 	all_caggs_info.max_bucket_widths = NIL;
 	all_caggs_info.mat_hypertable_ids = NIL;
+	all_caggs_info.bucket_functions = NIL;
+
 	Assert(list_length(caggs) > 0);
 
 	foreach (lc, caggs)
 	{
 		ContinuousAgg *cagg = lfirst(lc);
 
-		bucket_width = Int64GetDatum(ts_continuous_agg_bucket_width(cagg));
+		bucket_width = Int64GetDatum(ts_continuous_agg_bucket_width_variable(cagg) ?
+										 BUCKET_WIDTH_VARIABLE :
+										 ts_continuous_agg_bucket_width(cagg));
 		all_caggs_info.bucket_widths =
 			lappend(all_caggs_info.bucket_widths, DatumGetPointer(bucket_width));
 
-		bucket_width = Int64GetDatum(ts_continuous_agg_max_bucket_width(cagg));
+		bucket_width = Int64GetDatum(ts_continuous_agg_bucket_width_variable(cagg) ?
+										 BUCKET_WIDTH_VARIABLE :
+										 ts_continuous_agg_max_bucket_width(cagg));
 		all_caggs_info.max_bucket_widths =
 			lappend(all_caggs_info.max_bucket_widths, DatumGetPointer(bucket_width));
 
+		bucket_function = palloc(sizeof(*bucket_function));
+		*bucket_function = cagg->bucket_function;
+		all_caggs_info.bucket_functions = lappend(all_caggs_info.bucket_functions, bucket_function);
+
 		all_caggs_info.mat_hypertable_ids =
 			lappend_int(all_caggs_info.mat_hypertable_ids, cagg->data.mat_hypertable_id);
 	}
 	return all_caggs_info;
 }
 
+/*
+ * Serializes ContinuousAggsBucketFunction* into a palloc'ated string like:
+ *
+ *     ver;bucket_width;origin;timezone
+ *
+ * ... where ver is a version of the serialization format. This particular format
+ * was chosen because of it's simplicity and good performance. Future versions
+ * of the procedure can use other format (like key=value or JSON) and/or add
+ * extra fields. NULL pointer is serialized to an empty string.
+ *
+ * Note that the schema and the name of the function are not serialized. This
+ * is intentional since this information is currently not used for anything.
+ * We can serialize the name of the function as well when and if this would be
+ * necessary.
+ */
+static const char *
+bucket_function_serialize(const ContinuousAggsBucketFunction *bf)
+{
+	const int ver = 1;
+	StringInfo str;
+
+	if (bf == NULL)
+		return pstrdup("");
```

Comment:
Is there a reason we have to use `pstrdup()` here? Why is there a requirement for the return value to only be `palloc()`ed?

Reply:
Since the result is `const` anyway I have to agree that pstrdup() is redundant here.

Reply:
Fixed 👍